internal/pkg/flink/app/application.go

```diff
@@ -1,62 +1,165 @@
+		restoreStdin(stdinBefore)
+	})
 
 	// Instantiate Component Controllers
-	fetchController := controller.NewFetchController(dataStore)
-	tableController := controller.NewTableController(table, appController, fetchController, appOptions.GetVerbose())
-	inputController := controller.NewInputController(tableController, appController, dataStore, authenticated, historyStore, &appOptions)
+	inputController := controller.NewInputController(appController, historyStore)
+	statementController := controller.NewStatementController(appController, dataStore, consoleParser)
+	interactiveOutputController := controller.NewInteractiveOutputController(resultFetcher, appOptions.GetVerbose())
+	basicOutputController := controller.NewBasicOutputController(resultFetcher, inputController.GetWindowWidth)
 
-	// Pass RunInteractiveInputFunc to table controller so the user can come back from the output view
-	tableController.SetRunInteractiveInputCallback(inputController.RunInteractiveInput)
+	app := Application{
+		history:                     historyStore,
+		store:                       dataStore,
+		resultFetcher:               resultFetcher,
+		appController:               appController,
+		inputController:             inputController,
+		statementController:         statementController,
+		interactiveOutputController: interactiveOutputController,
+		basicOutputController:       basicOutputController,
+		tokenRefreshFunc:            synchronizedTokenRefresh(tokenRefreshFunc),
+	}
+	components.PrintWelcomeHeader()
+	app.readEvalPrintLoop()
+}
 
-	// Event handlers
-	app.SetInputCapture(tableController.AppInputCapture)
+func getStdin() *term.State {
+	state, err := term.GetState(int(os.Stdin.Fd()))
+	if err != nil {
+		log.CliLogger.Warnf("Couldn't get stdin state with term.GetState. Error: %v\n", err)
+		return nil
+	}
+	return state
+}
 
-	interactiveOutput := components.InteractiveOutput(table, shortcuts)
-	rootLayout := components.RootLayout(interactiveOutput)
+func getConsoleParser() prompt.ConsoleParser {
+	consoleParser := prompt.NewStandardInputParser()
+	err := consoleParser.Setup()
+	if err != nil {
+		log.CliLogger.Warnf("Couldn't setup console parser. Error: %v\n", err)
+	}
+	return consoleParser
+}
+
+func tearDownConsoleParser(consoleParser prompt.ConsoleParser) {
+	err := consoleParser.TearDown()
+	if err != nil {
+		log.CliLogger.Warnf("Couldn't tear down console parser. Error: %v\n", err)
+	}
+}
 
-	// We start tview and then suspend it immediately so we initialize all components
-	app.SetAfterDrawFunc(func(screen tcell.Screen) {
-		if !screen.HasPendingEvent() {
-			once.Do(func() {
-				go appController.SuspendOutputMode(inputController.RunInteractiveInput)
-			})
+func restoreStdin(state *term.State) {
+	if state != nil {
+		_ = term.Restore(int(os.Stdin.Fd()), state)
+	}
+}
+
+func (a *Application) readEvalPrintLoop() {
+	for a.isAuthenticated() {
+		userInput := a.inputController.GetUserInput()
+		if a.inputController.IsSpecialInput(userInput) {
```

Comment:
I don't love the name `isSpecialInput` but don't have a better suggestion. But we should add some comments of top of the function since its name doesn't say a lot.

Reply:
Hmmm in that case I'll try to think of a better name, because I think a better name would make the comment unnecessary!

Reply:
split up the method now, which should make it a lot clearer what's happening 👍